### ISSUE TITLE
stability: 30-day timeseries view needs 3GB ram

### ISSUE THREAD

**Issue by danhhz** [cockroachdb/cockroach#20014]
While attempting to bless the release scheduled to go out today, adriatic node 5 OOM'd. The last message in the logs from server/status/runtime.go:219:

```
I171113 16:06:09.626277 298 server/status/runtime.go:219  [n5] runtime stats: 3.5 GiB RSS, 264 goroutines, 578 MiB/193 MiB/917 MiB GO alloc/idle/total, 2.0 GiB/2.5 GiB CGO alloc/total, 3259.00cgo/sec, 1.02/0.35 %(u/s)time, 0.00 %gc (1x)
```

oom killer:
```
[Mon Nov 13 16:06:52 2017] Killed process 15895 (cockroach) total-vm:13056228kB, anon-rss:7756876kB, file-rss:0kB
```

When I brought the node back up, it happened again:
```
I171113 16:26:48.259964 262 server/status/runtime.go:219  [n5] runtime stats: 2.9 GiB RSS, 260 goroutines, 98 MiB/286 MiB/431 MiB GO alloc/idle/total, 1.9 GiB/2.4 GiB CGO alloc/total, 2652.68cgo/sec, 0.90/0.37 %(u/s)time, 0.00 %gc (6x)
```

```
[Mon Nov 13 16:27:25 2017] Killed process 21916 (cockroach) total-vm:9718100kB, anon-rss:7767344kB, file-rss:0kB
```

It was running the normal kv load generator. Nothing had changed in a while except that I was clicking around the admin ui graphs (and loading the ones going back a month), unsure if that's related. I've left node 5 down, so it may potentially repro by restarting it.

**Comment by danhhz** [comment]
I'm a little out of my depth here, @bdarnell can you look at this or reassign it to someone who can?

**Comment by bdarnell** [comment]
The "disagreement" is easy to explain: the runtime.go logging is a sample once every 10 seconds; the OOM killer sees the actual values at the time of the kill. That just means that whatever ate up all the memory grew very quickly. 

Retitled to reflect that the problem is something consuming gigabytes of ram in a few seconds instead of a monitoring issue.

**Comment by mberhault** [comment]
The memstats are ~45s before the OOM.
Checking monitoring (polls every 10s), we can see a big climb in Go memory usage just before:
![go-mem](https://user-images.githubusercontent.com/5901480/32740915-7af53b76-c872-11e7-8094-aff3bae16a8e.png)

**Comment by bdarnell** [comment]
In the time between the last stats logging and the (second) OOM, the logs show that the node

- processed a number of preemptive snapshots (the largest was 32MB)
- had some slow rocksdb commits (719ms while applying a snapshot) and (probably related) slow handleRaftReady calls
- added about 20 replicas to the replica gc queue (peer suggestion)
- got some new range leases

Nothing really stands out as unusual here except that there seems to be a lot of rebalancing going on.

On the second OOM, the stats logger showed a jump in go memory usage:

```
I171113 16:27:08.260317 262 server/status/runtime.go:219  [n5] runtime stats: 2.9 GiB RSS, 345 goroutines, 268 MiB/119 MiB/431 MiB GO alloc/idle/total, 2.0 GiB/2.4 GiB CGO alloc/total, 1267.50cgo/sec, 0.44/0.18 %(u/s)time, 0.00 %gc (4x)
I171113 16:27:18.450380 262 server/status/runtime.go:219  [n5] runtime stats: 4.5 GiB RSS, 496 goroutines, 2.0 GiB/85 MiB/2.2 GiB GO alloc/idle/total, 2.0 GiB/2.4 GiB CGO alloc/total, 1042.18cgo/sec, 0.89/0.44 %(u/s)time, 0.00 %gc (11x)
```

I'll see if the profiles show anything.

**Comment by bdarnell** [comment]
![image](https://user-images.githubusercontent.com/160562/32742993-115bdd1c-c879-11e7-9668-709ed33a0300.png)

Omega 4 had a similar burst of go memory usage, but didn't crash because it's running on a machine with more memory. Profiles from this time period have been deleted. 

Since this happened at the same time @danhhz was poking around the UI, I suspect it's related to that. Maybe the 1-month UI view loads too much data, or maybe it was slow enough to load that it timed out and retried, loading more copies of the data on the server side.

**Comment by bdarnell** [comment]
Assigning to @mrtracy as the expert on our timeseries subsystem. Ideally timeseries data should be subject to some sort of memory monitoring/limiting (like we do with SQL memory usage). We could also reduce the number of time series fetched per request from the UI, or just remove the 30-day option as a quick fix.

**Comment by mrtracy** [comment]
Mitigations I can think of, in order of complexity to implement

+ Remove the 30 day window option.
+ Reduce the absolute number of TS queries a node can access in parallel.  This is currently enforced by a "LimitedAsyncQuery" in the time series server, and currently we allow 250 parallel queries. We could reduce that number, and that could mitigate this problem if it is the result of numerous parallel queries - this does not solve the problem if *one* query is taking up too much memory, and it could slow down queries over smaller timespans.
+ Rate limit queries on the front end. For example, we can test if a graph is currently visible on-screen and only query time series in that case; we could also query graphs one at a time, rather than batching the requests together like we currently do.
+ Memory management - attempt to install some sort of memory monitoring/limiting in analogy to the SQL system, and slow down queries if we are going over that limit.  This would be a more advanced version to the "LimitedAsyncQuery" solution, which would still allow a large number of parallel queries for smaller timespans.  However, in the case where a single query is taking up a large amount of memory, this still might not solve the problem.
+ Downsampling on disk for old time series - that is, instead of deleting all time series after 30 days, we permanently downsample them after a shorter time period (say, one week).  This would result in lower fidelity data being available for older timespans, but that would not matter when viewing a very large timespan; and it would have the effect of dramatically lowering the amount of data needed in memory to service a large timespan query.  This has a straightforward implementation for downsampling data, but would require non-trivial work to support it in queries.
+ "Streaming" data during queries; currently, we query all data needed to complete a single query into memory.  However, this could naturally be modified into a streaming system, which would dramatically reduce the amount of memory needed at any one time for a single large query.

**Comment by petermattis** [comment]
@mrtracy It isn't clear from your message: are these options being suggested after looking at a heap profile of timeseries query memory usage? 

> "Streaming" data during queries; currently, we query all data needed to complete a single query into memory. However, this could naturally be modified into a streaming system, which would dramatically reduce the amount of memory needed at any one time for a single large query.

This seems like a good option, though I'd like to see the analysis of a heap profile during a timeseries query to verify it will address the problem.

**Comment by spencerkimball** [comment]
Yes streaming sounds like an excellent direction.

Longer term seems we should also explore downsing.

**Comment by mrtracy** [comment]
No, those suggestions are based on my intuition of what would be using memory in this system.  I suppose there could be something unexpected at play, but if a single request queries a large number of time series over a large timespan, using a large chunk of memory wouldn't surprise me at all.  In order to service a large timespan query, we currently pull *all* data for that a time series over that timespan into memory at once; that's the only appreciable source of memory, as the result is downsampled considerably and the iterators are very efficient.

Also, note that I ordered those suggestions by difficulty of implementation: I think streaming would be the most difficult of all of those options to implement, due to complexity of both implementation and testing.

**Comment by petermattis** [comment]
> No, those suggestions are based on my intuition of what would be using memory in this system. 

"Trust, but verify".

> Also, note that I ordered those suggestions by difficulty of implementation: I think streaming would be the most difficult of all of those options to implement, due to complexity of both implementation and testing.

Seems unlikely that "streaming" which only affects queries will be more difficult than downsampling which affects writing. Just my opinion.

**Comment by mrtracy** [comment]
The downsampling to disk is straightforward because it is very similar to the culling procedure, which is already implemented (in fact, this has already gotten RFC treatment). Where it becomes difficult is querying over data with multiple on-disk resolutions.

The streaming is not straightforward because it requires a refactoring of both querying and the processing iterator.

However, it's been a while since I thought about this; maybe my brain is confusing the perceived difficulty of the streaming refactors with the perceived difficulty of querying mixed-resolution data from disk.

**Comment by dianasaur323** [comment]
Just spoke to @mrtracy off-line. It sounds like we might need 2 different approaches here - one that can fix the OOM stability issue in the short term, and another to actually build a solution that fixes this problem. Some suggestions that have been proposed (here and elsewhere) are to turn of the 30 day option, or to lower that option to something like 14 days. We won't have time to build out anything like streaming in this current development cycle given the pressure we already have to ship some other large features. Could we find a solution that solves the OOM problem in the short term for now, and revisit how to make this better in the next release cycle?

**Comment by petermattis** [comment]
@mrtracy What are you imagining the "streaming" approach would look like? Changing `DB.Query` to limit the number of keys retrieved at a time and then down-sampling the data in memory seems feasible.

**Comment by mrtracy** [comment]
The downsampling iteration process is going to have to become asynchronous, and the data container is now going to have to track where the iterator is located in order to query additional data (and release data that is no longer needed).  This is a *significant* refactoring of that entire algorithm.

**Comment by cuongdo** [comment]
Reassigning to @solongordon to be done after his other tasks

**Comment by mrtracy** [comment]
Math for this:

When servicing a typical query, the current system will load all data for the time series being queried for all nodes in the time span being queried.

+ A single time series sample is *40* bytes.
+ Each hour of time series data is 40 bytes (for a container) + 360 samples =~ *14.4k*
+ For a 30 days, this is 14.4k * 24hours * 30days =~ *10.8M* for 30 days of one time series for one node.
+ For a six node cluster, this is ~62M of data.

That's a lower bound of the amount of memory needed, at one time, to service one query at the 30 day timespan on a six node cluster.

The other compounding factor is the number of queries being executed at one time; the dashboard with the most active queries at once is the queues dashboard, which will dispatch 37 queries at once. On a six node cluster, this needs about 2.3G of memory on a six node cluster.  Although these queries are executed independently, they will be rapidly dispatched by the time series server and there is no memory-related pressure applied which would prevent them all from running simultaneously.

Some visual evidence of this:

<img width="1336" alt="screen shot 2017-12-13 at 7 02 34 pm" src="https://user-images.githubusercontent.com/6969858/33968952-c75d92dc-e038-11e7-8f10-40afa77de17c.png">

The above is me pulling up the 1-month view on Cobalt, a six node cluster.  It has not been running for a full month, but it has been running for almost exactly ten days (240 hours).  For a six node cluster, we would predict that this view would use 14.4 * 24 * 10 * 37 * 6 =~ *770M*

<img width="842" alt="screen shot 2017-12-13 at 7 18 53 pm" src="https://user-images.githubusercontent.com/6969858/33969299-7fbeecc6-e03a-11e7-8f0e-b8a4537a6808.png">

Viewing this correlated with a 800-900MB spike on a single machine (both of the allocated spikes in the above example correlate exactly with my viewing of the queues dashboard).

This visual evidence, along with our knowledge of the query system, is enough for me to conclude that this is indeed the root cause.

As for the solution: we need to cut down on the amount of time series data in memory at one time.  The data grows in three dimensions:

+ Time
+ Nodes
+ Simultaneous Queries

Simultaneous queries does not seem like an effective answer, as on a 100 node cluster a single 30 day query would require a gigabyte of data in memory.

So we are left with "time" and "nodes".

+ Time would involve "streaming" the time series data into memory, with only a slice of the full time span in memory at each time.  As the downsampling iterator processes the data, new data would be queried while old data would be released. This would allow us much better control over the amount of data in memory.
+ To reduce the contribution of nodes, you would limit each .  Each "query" would be limited to data for a single node, and you would combine the results of multiple queries into the returned value.

I think that the "time" option is best for the long term, primarily because of the way that time series keys are laid out in memory.  Time series data is sorted by (series name, hour, node), meaning that data for one node at a particular series and time is immediately adjacent to the same data for all other nodes.  Therefore, splitting queries into smaller time spans is much more efficient than splitting queries by node.

**Comment by petermattis** [comment]
Thanks for the analysis, @mrtracy.

**Comment by mrtracy** [comment]
This is a bit complicated to understand, so some additional information on where this is headed:

"Streaming" data is fairly complicated because of the fact that we *interpolate* datapoints and *aggregate* data points.

+ Interpolation is used when a data point is missing for a particular time series.  This is usually fairly benign (i.e. due to clock oddness, one ten-second sample is randomly missing from a node), but can result from actual node downtime, and it's difficult to tell the difference.
+ Aggregation involves combining the values of multiple time series together into a single output. Right now, this is only used to combine the same named time series (e.g. "sql.inserts.count") as recorded from different nodes into a single aggregate for the cluster.

Interpolation makes it difficult to stream data because, at the boundaries of our one-hour data slabs, we may need to interpolate across slabs; that is, in order to compute a missing value at the end of an hour, we need data from the next hour. This increases the difficulty of streaming considerably.

Aggregation is even more difficult, because of the fact that nodes can be added or removed from the cluster.  For a very long time period (e.g. a month), it's likely that several new nodes will be brought online, or that several nodes will be decommissioned, or even that nodes will be temporarily offline but reappear later.  It is challenging to create an efficient abstraction that aggregates across multiple data streams where streams can appear and disappear.

To simplify things a bit and work around these difficulties, we will introduce an *interpolation limit*: a maximum time span over which data can be interpolated. This will be set to the cluster setting used to determine that a node is down; if a gap appears in a time series that is longer than this, the intermediate points will no longer be interpolated and no data will be returned.

The benefit is that this lets us get a correct result while having a deterministic limit on how much data needs to be in memory at one time: we can fully interpolate and aggregate the data for any time span with only one additional "slab" of data before and after that span.  This greatly reduces the needed refactoring to fix this bug.

The downside occurs when missing data and aggregation intersect; this is effect is already visible in #13537, which occurs on the leading edge of data.  With interpolation limits, a short downtime for a node would result in sudden dips in aggregated time series; however, I believe that this downside will be fully ameliorated once events on graphs is (finally) added, which will allow us to correlate these dips with the fact that nodes were down.

**Comment by mrtracy** [comment]
Current plan of action, outline from a PR:

The overall plan is to have resource governance before using the memory management, using the following strategy:
+ Compute a "soft-limit" of memory for use by the time series query system, and reserve it.
+ Limit the number of simultaneous workers servicing queries (currently 250, might be adjusted).
+ Divide the soft limit evenly by this number to determine the budget per-worker.
+ Queries for large time spans will be broken up among multiple workers if the amount of memory needed for the whole query would exceed the per-worker budget.
+ Each worker will reserve some space from the soft-limit by estimating how much memory it will need. This estimation will include the total number of active nodes on the cluster.
+ The workers will use still use the "Grow" mechanism to catch places where the estimation is somehow wrong and abort queries if necessary.

Essentially, the system is designed such that:
+ Ideally, no query should ever be aborted due to memory constraints due to the first-level resource governance
+ If thats resource governance fails, queries will be aborted by the memory management system, which will be noticeable in our logs.
